homeassistant/components/leviosa_shades/cover.py

```diff
@@ -0,0 +1,180 @@
+"""The Leviosa Shades Zone base entity."""
+import logging
+
+from leviosapy import LeviosaShadeGroup as tShadeGroup, LeviosaZoneHub as tZoneHub
+import voluptuous as vol
+
+from homeassistant.components import cover
+from homeassistant.const import ATTR_ENTITY_ID
+from homeassistant.core import HomeAssistant
+from homeassistant.helpers import config_validation as cv, entity_platform
+from homeassistant.helpers.aiohttp_client import async_get_clientsession
+
+from .const import (
+    BLIND_GROUPS,
+    DOMAIN,
+    MANUFACTURER,
+    MODEL,
+    SERVICE_NEXT_DOWN_POS,
+    SERVICE_NEXT_UP_POS,
+)
+
+_LOGGER = logging.getLogger(__name__)
+
+# Estimated time it takes to complete a transition
+# from one state to another
+TRANSITION_COMPLETE_DURATION = 30
+PARALLEL_UPDATES = 1
+
+COVER_NEXT_POS_SCHEMA = vol.Schema({vol.Optional(ATTR_ENTITY_ID): cv.entity_ids})
+
+
+async def async_setup_entry(hass, entry, async_add_entities):
+    """Set up the Leviosa shade groups."""
+    _LOGGER.debug(
+        "Setting up %s[%s]: %s",
+        entry.domain,
+        entry.title,
+        entry.entry_id,
+    )
+    hub_name = entry.title
+    hub_mac = entry.data["device_mac"]
+    hub_ip = entry.data["host"]
+    blind_groups = entry.data[BLIND_GROUPS]
+    _LOGGER.debug("Groups to create: %s", blind_groups)
+    hub = tZoneHub(
+        hub_ip=hub_ip, hub_name=hub_name, websession=async_get_clientsession(hass)
+    )
+    await hub.getHubInfo()  # Check all is good
+    _LOGGER.debug("Hub object created, FW: %s", hub.fwVer)
+    entities = []
+    for blind_group in blind_groups:
+        _LOGGER.debug("Adding blind_group: %s", blind_group)
+        new_group_obj = hub.AddGroup(blind_group)
+        entities.append(
+            LeviosaBlindGroup(
+                hass, hub_mac + "-" + str(new_group_obj.number), new_group_obj
+            )
+        )
+    async_add_entities(entities)
```

Comment:
Please move this to your __init__.py. The logic for connecting should not be part of your cover.py.

See https://github.com/home-assistant/core/blob/dev/homeassistant/components/overkiz/__init__.py (and many other integration) for an example.

Reply:
Thank you for your suggestion.
My understanding is that since a leviosa shades can spawn multiple entities, this curtain controller is essentially a wifi to IR device, but it can control up to six groups of covers, so it will derive up to seven entities, such as `group1` - `group6` and `group_ALL`, so we need to call the `async_add_entities` method inside the `async_setup_entry` method, but obviously the `async_setup_entry` method in `__init__.py` doesn't have an `AddEntitiesCallback` parameter, so it's not possible to call `async_add_entities` inside `async_setup_entry` in __init__,py anymore.
I'm currently stuck at this point and don't know how to proceed from here, again I'll ask you for your advice, thanks!

Reply:
By the way, I looked at your sample code and I see that your scenario is logging into a cloud account and then being able to register multiple `devices`, not `entities`, And the situation I encountered is that one `device` supports multiple `entities`

Reply:
I refer to: https://github.com/home-assistant/core/blob/7d45227a5966bec0d265026dd36bd83df7e46cff/homeassistant/components/acmeda/cover.py#L22C13-L22C13

Reply:
@greg-ha-1990 in the example you link to, the hub is instantiated in `__init__.py` and retrieved via the `hass.data`. This is the same as my suggestion.

````python3
    hub: PulseHub = hass.data[DOMAIN][config_entry.entry_id]
```

Please don't resolve conversations that you didn't fix. This makes it hard for reviewers to look at the open issues.

Reply:
Yes. And in `__init__` you should check if you can connect, and otherwise catch the error / raise ConfigEntryNotReady (for retry) or another status.

Cover.py should only be responsible for setting up the entity, not all the connections around this.

Reply:
Okay, I have updated the code and the test passes, but I don’t know whether to click `Resolve conversation` at this time, so I won’t click this button yet.

Reply:
You didn't push your changes yet?

In essence, it is fine to use 'Resolve conversation' if you are sure this issue is fixed. If you aren't sure, just leave it open and a reviewer will close it when applicable. If you just close reviews you don't agree with / have different opinions on, it is hard to keep track of the discussion and open feedback.

Reply:
I just pushed the code